src/Middleware/RequestDecompression/src/DeflateDecompressionProvider.cs

```diff
@@ -13,6 +14,221 @@ internal sealed class DeflateDecompressionProvider : IDecompressionProvider
+        public override long Position { get => _stream.Position; set => _stream.Position = value; }
+
+        public override void Write(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+        public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken) => throw new NotSupportedException();
+
+        // On the first read request, peek at the first nibble of the response. If it's an 8, use ZLibStream, otherwise
+        // use DeflateStream. This heuristic works because we're deciding only between raw deflate and zlib wrapped around
+        // deflate, in which case the first nibble will always be 8 for zlib and never be 8 for deflate.
+        // https://stackoverflow.com/a/37528114 provides an explanation for why.
+
+        public override int Read(Span<byte> buffer)
+        {
+            if (_decompressionStream is null)
+            {
+                int firstByte = _stream.PeekFirstByte();
+                _decompressionStream = CreateDecompressionStream(firstByte, _stream);
+            }
+
+            return _decompressionStream.Read(buffer);
+        }
+
+        public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            if (_decompressionStream is null)
+            {
+                return CreateAndReadAsync(this, buffer, cancellationToken);
+
+                static async ValueTask<int> CreateAndReadAsync(ZLibOrDeflateStream thisRef, Memory<byte> buffer, CancellationToken cancellationToken)
+                {
+                    int firstByte = await thisRef._stream.PeekFirstByteAsync(cancellationToken).ConfigureAwait(false);
+                    thisRef._decompressionStream = CreateDecompressionStream(firstByte, thisRef._stream);
+                    return await thisRef._decompressionStream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return _decompressionStream.ReadAsync(buffer, cancellationToken);
+        }
+
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            ValidateCopyToArguments(destination, bufferSize);
+            return Core(destination, bufferSize, cancellationToken);
+            async Task Core(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                if (_decompressionStream is null)
+                {
+                    int firstByte = await _stream.PeekFirstByteAsync(cancellationToken).ConfigureAwait(false);
+                    _decompressionStream = CreateDecompressionStream(firstByte, _stream);
+                }
+
+                await _decompressionStream.CopyToAsync(destination, bufferSize, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count) => _stream.Read(buffer, offset, count);
```

Comment:
this is missing the peeking.

Reply:
I guess this is the same thing as saying -- every operation on _stream should be peeking to set up _decompressionStream then operation on _decompressionStream.

Reply:
Note that the HttpClient version derives from an internal base stream that inverts the dependencies been the various methods and then it only needs to override the span and memory based reads.